### PR TITLE
feat(Cassandra): add Cassandra implementation of EntityService

### PR DESCRIPTION
**Pull request by xdl**
This PR implements a new EntityService (DatastaxEntityService) and its associated DAO, aspect classes etc. in order to support Datastax's [Cassandra](https://en.wikipedia.org/wiki/Apache_Cassandra) for DataHub's [document store](https://datahubproject.io/docs/architecture/metadata-serving#metadata-query-serving) component.

The implementation (Cassandra Datastax or Ebean SQL) can be set with the `DAO_SERVICE_LAYER` environment variable (e.g. in `/docker/datahub-gms/env/docker.cassandra.env`). The other corresponding DB configs that can be set are:

* `DATASTAX_DATASOURCE_USERNAME` (e.g. `cassandra`)
* `DATASTAX_DATASOURCE_PASSWORD` (e.g. `cassandra`)
* `DATASTAX_HOSTS` (e.g. `localhost`)
* `DATASTAX_PORT` (e.g. `9042`)
* `DATASTAX_DATACENTER` (e.g. `datacenter1`)
* `DATASTAX_KEYSPACE` (e.g. `datahubpoc`)
* `DATASTAX_USE_SSL` (e.g. `false`)

To test out locally, do a build and then run `./docker/dev-with-cassandra.sh`.

## Motivation

While evaluating DataHub for internal use, we identified that it was important for the query serving to be able to be horizontally scalable, as we've come up with scaling issues for one of our use cases relying on a traditional SQL store that DataHub would replace. Along with the requirement that the database should be highly available and with native support for multi DC replication, this led us to evaluate the use of Cassandra, a horizontally scalable database as part of our data catalog solution.

### Overview

* We use the Cassandra library from [testcontainers](https://www.testcontainers.org/modules/databases/cassandra/) to provide an out-of-process Cassandra instance for integration tests (`DatastaxEntityServiceTest.java`)
* The architecture (and tests) is inspired from the `EbeanEntityService` and the other Ebean implementations
* Some nuances to queries, inserts and updates in a distributed database (e.g. see Paging and Sorting), are discussed below

### Conditional Updates vs Transactions

Cassandra does not have ACID transactions; it does however have [conditional updates](https://docs.datastax.com/en/developer/java-driver/4.4/manual/mapper/daos/update/) which we make use of in `DatastaxAspectDao.batchSaveLatestAspect` and `DatastaxAspectDao.condUpsertAspect` to avoid a potential race condition when ingesting new entities (see https://github.com/linkedin/datahub/issues/3231).

### SCSI Support

Since Cassandra does not support `JOIN`s, it would be difficult to implement SCSI without adding more columns into the `metadata_aspect` table, although it's our understanding that if this feature is about to be dropped, this wouldn't be a long-term issue.

### Paging and Sorting

This mostly relates to the reimplementing of the `listLatestAspects` method. A new column in the DB has been added for the entity type (e.g. `corpuser`), populated upon insertion. The alternative to this would be to use a [SASI](https://docs.datastax.com/en/dse/5.1/cql/cql/cql_using/useSASIIndex.html), although there are some sources suggesting that it won't be any more performant (https://www.scnsoft.com/blog/cassandra-performance).

Furthermore, the sorting order (ascending by URN) feature has been dropped as it would have to be done in memory (since URN is used as the partition key), and this condition has been relaxed in `testListLatestAspects`.

Cassandra's paging has been implemented using the [OffsetPager](https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/paging/); there does not seem to be a performant way of getting the total number of matches back (e.g. https://www.datastax.com/blog/running-count-expensive-cassandra). Our use case isn't heavily focused on the frontend so there hasn't been much effort expended in optimising this.

### Anticipated Further Work

* Update to use DAO (https://docs.datastax.com/en/developer/java-driver/4.4/manual/mapper/daos/update/) class annotations which would be more idiomatic than the current implementation of the DatastaxAspect class
* Implement and update to use `EntityService.ingestProposal`
* Further investigation into rolling back (deleting) aspects
* Optimisation around the `batchGet` method
* Optimisation around frontend (`listLatestAspects`, `listUrns`) use cases

### CODE REVIEW

File: metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+
+public class DatastaxEntityServiceTest {
```
Comment (EnricoMi) [inline metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java:62]: This file is very similar to `EbeanEntityServiceTest`. The tests and common helper methods should be moved into a base test class and only implementation specific tests and code should be in these test files.

**Comment by EnricoMi** [comment]
@xdl is this limited to Datastax Cassandra, or does it perfectly work with Apache Cassandra?

**Comment by xdl** [comment]
Thanks for the feedback @jjoyce0510 and @EnricoMi  (particularly with the test speedup)! I've addressed some of the feedback; I think further actions from my end, and it'd be good to make sure we're on the same page before actioning, are:

* Rename Datastax* to Cassandra* since as pointed out by @EnricoMi we're only using the datastax libraries, but still Apache Cassandra
* Refactor common business logic from Ebean/C* back into EntityService (by fleshing out the EntityAspect interface more) and try to keep most of the implementation differences in the DAOs.

In the meantime the `dev-with-cassandra.sh` script should be working so if there are any feedback regarding usage that'd be great.

Another question that came up as I was bringing master up to date: I noticed the inclusion of a `IngestDataPlatformInstancesStep.java` bootstrap script that populates the DB with data platform instance aspects which I've yet to implement on the Cassandra side; can you elaborate what this is doing and what the implications of leaving it off are, as the data platforms themselves are being populated from a hardcoded `data_platforms.json` file, and then the `IngestDataPlatformInstancesStep` seems to be intended to run just once after that (it'll stop running once there's a dataplatforminstance existing in the DB)?

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/DatastaxEntityService.java [hunk 2]
```diff
@@ -0,0 +1,526 @@
+package com.linkedin.metadata.entity.datastax;
+
+import com.datastax.oss.driver.api.core.cql.ResultSet;
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/DatastaxEntityService.java:4]: Yeah agree. Any lower hanging fruit you see, feel free to address!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/DatastaxEntityService.java [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+  public void setAlwaysEmitAuditEvent(Boolean alwaysEmitAuditEvent) {
+    _alwaysEmitAuditEvent = alwaysEmitAuditEvent;
+  }
+
+  public void setWritable(boolean canWrite) {
+  }
+
+  @Override
+  public Urn ingestProposal(@Nonnull MetadataChangeProposal metadataChangeProposal, AuditStamp auditStamp) {
+    return null;
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/DatastaxEntityService.java]: Oh that'd be great! 

We are currently in the process of migrating _to_ ingestionProposal, so we will need that one to be implemented :)

File: metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+
+public class DatastaxEntityServiceTest extends EntityServiceTestBase {
+
+  private CassandraContainer _cassandraContainer;
+
+  @BeforeMethod
+  public void setupTest() {
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java]: Yes!

File: metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+    final DockerImageName imageName = DockerImageName
+            .parse("cassandra:3.11")
+            .asCompatibleSubstituteFor("cassandra");
+
+    _cassandraContainer = new CassandraContainer(imageName);
+    _cassandraContainer.start();
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/test/java/com/linkedin/metadata/entity/DatastaxEntityServiceTest.java:52]: Thank you!!

**Comment by jjoyce0510** [comment]
> Another question that came up as I was bringing master up to date: I noticed the inclusion of a IngestDataPlatformInstancesStep.java bootstrap script that populates the DB with data platform instance aspects which I've yet to implement on the Cassandra side; can you elaborate what this is doing and what the implications of leaving it off are, as the data platforms themselves are being populated from a hardcoded data_platforms.json file, and then the IngestDataPlatformInstancesStep seems to be intended to run just once after that (it'll stop running once there's a dataplatforminstance existing in the DB)?

Your understanding here is correct. It simply serves to pick up your set of data_platforms from that file and get them into DataHub on first deploy. 

If you omit this step, in general you will be okay with or without ingesting them. However, to see the platform logos in the UI you'll need to ingest them. You can alternatively ingest from a different file, `mce_files/data_platforms.json` using the datahub CLI

**Comment by jjoyce0510** [comment]
@xdl And the remaining items look good to me!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityAspect.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.linkedin.metadata.entity;
+
+public interface EntityAspect {
+    String getSystemMetadata();
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/EntityAspect.java:4]: Sounds good to me!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.linkedin.metadata.entity.datastax;
+
+public class ConditionalWriteFailedException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java:3]: Based on the explanation I think exception doesn't hurt if the caller will benefit by having the context about _why_ the operation failed.. feel free to close

**Comment by jjoyce0510** [comment]
@xdl ! How are things going?

**Comment by jjoyce0510** [comment]
@xdl Checking in again. Do you need help revisiting this?

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (RyanHolstien) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java:3]: Agree with Enrico here, this looks like it can just be a boolean essentially based on the usage. Exception throwing is heftier than is needed here.

**Review by RyanHolstien — APPROVED** [review]
Overall, LGTM. A couple of minor comments :) Thanks so much for all the hard work put into this PR!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java (hunk 7 above)
Comment (EnricoMi) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java:3]: It does not look like this exception is used anywhere.

**Review by EnricoMi — CHANGES_REQUESTED** [review]
The rebranding from `Datastax` to `Cassandra` is still missing in some places (class names): https://github.com/linkedin/datahub/pull/3286/files#r732708283

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/DatastaxAspectDaoFactory.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.context.annotation.DependsOn;
+
+import javax.annotation.Nonnull;
+
+
+@Configuration
+public class DatastaxAspectDaoFactory {
+
+  @Autowired
+  ApplicationContext applicationContext;
```
Comment (RyanHolstien) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/DatastaxAspectDaoFactory.java:19]: Was this causing issues? Directly using the ApplicationContext to retrieve beans is generally frowned upon. Really this shouldn't even need the @Autowired, you should be able to use method level injection i.e.:

protected DatastaxAspectDao createInstance(CqlSession cqlSession) {...}

File: metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java (hunk 7 above)
Comment (justinas-marozas) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/datastax/ConditionalWriteFailedException.java]: Removed

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/DatastaxAspectDaoFactory.java (hunk 8 above)
Comment (justinas-marozas) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/DatastaxAspectDaoFactory.java]: Good call. Done

File: docker/docker-compose.m1.yml [hunk 9]
```diff
@@ -6,4 +6,6 @@ services:
   schema-registry:
     image: eugenetea/schema-registry-arm64:latest
   zookeeper:
-    image: kymeric/cp-zookeeper:latest
\ No newline at end of file
+    image: kymeric/cp-zookeeper:latest
+  neo4j:
```
Comment (EnricoMi) [inline docker/docker-compose.m1.yml:10]: how is this change related to a "Cassandra implementation of EntityService"?

**Review by EnricoMi — COMMENTED** [review]
LGTM, just one minor question on the added neo4j dependency.

**Review by RyanHolstien — APPROVED** [review]
Changes since last review lgtm

File: metadata-io/src/test/java/com/linkedin/metadata/entity/EbeanEntityServiceTest.java [hunk 10]
```diff
@@ -63,31 +31,20 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
+
+
+public class EbeanEntityServiceTest extends EntityServiceTestBase<EbeanAspectDao, EbeanEntityService, EbeanRetentionService> {
```
Comment (jjoyce0510) [inline metadata-io/src/test/java/com/linkedin/metadata/entity/EbeanEntityServiceTest.java:47]: Amazing!

File: metadata-io/src/test/java/com/linkedin/metadata/graph/dgraph/DgraphGraphServiceTest.java [hunk 11]
```diff
@@ -52,7 +52,7 @@ protected Duration getTestConcurrentOpTimeout() {
         return Duration.ofMinutes(5);
     }
 
-    @BeforeTest
+    @BeforeClass
```
Comment (jjoyce0510) [inline metadata-io/src/test/java/com/linkedin/metadata/graph/dgraph/DgraphGraphServiceTest.java:55]: THANK YOU!!

File: metadata-io/src/test/java/com/linkedin/metadata/search/LineageSearchServiceTest.java [hunk 12]
```diff
@@ -63,13 +64,13 @@ public class LineageSearchServiceTest {
   private static final String ENTITY_NAME = "testEntity";
   private static final Urn TEST_URN = TestEntityUtil.getTestEntityUrn();
 
-  @BeforeTest
+  @BeforeClass
```
Comment (jjoyce0510) [inline metadata-io/src/test/java/com/linkedin/metadata/search/LineageSearchServiceTest.java:67]: Wow - I am surprised to see we were previously doing this. No longer these tests felt so painful

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EntityServiceFactory.java [hunk 13]
```diff
@@ -3,35 +3,50 @@
...
+
 
 @Configuration
 public class EntityServiceFactory {
-  @Autowired
-  ApplicationContext applicationContext;
 
   @Bean(name = "entityService")
-  @DependsOn({"ebeanAspectDao", "kafkaEventProducer", TopicConventionFactory.TOPIC_CONVENTION_BEAN, "entityRegistry"})
+  @DependsOn({"cassandraAspectDao", "kafkaEventProducer", TopicConventionFactory.TOPIC_CONVENTION_BEAN, "entityRegistry"})
```
Comment (jjoyce0510) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EntityServiceFactory.java:26]: Awesome. Same comment about the environment variable- I'd prefer if we could use the YAML

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/RetentionServiceFactory.java [hunk 14]
```diff
@@ -23,19 +27,27 @@ public class RetentionServiceFactory {
...
   @Value("${RETENTION_APPLICATION_BATCH_SIZE:1000}")
   private Integer _batchSize;
 
 
+  @Bean(name = "retentionService")
+  @DependsOn({"cassandraSession", "entityService"})
+  @ConditionalOnProperty(name = "ENTITY_SERVICE_IMPL", havingValue = "cassandra")
+  @Nonnull
+  protected RetentionService createCassandraInstance(CqlSession session) {
+    RetentionService retentionService = new CassandraRetentionService(_entityService, session, _batchSize);
```
Comment (jjoyce0510) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/RetentionServiceFactory.java:39]: Huge improvement to the retention service, to decouple from Ebean. Awesome work.

**Review by jjoyce0510 — COMMENTED** [review]
Few final minor comments. Overall this looks ready to ship. Incredible work & greatly appreciate your patience with a fast-moving code base like ours

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java [hunk 15]
```diff
@@ -2,23 +2,22 @@
...
+
 
 @Configuration
 public class EbeanAspectDaoFactory {
-  @Autowired
-  ApplicationContext applicationContext;
 
   @Bean(name = "ebeanAspectDao")
   @DependsOn({"gmsEbeanServiceConfig"})
+  @ConditionalOnProperty(name = "ENTITY_SERVICE_IMPL", havingValue = "ebean", matchIfMissing = true)
```
[the remark this answers is not in the document]
Comment (RyanHolstien) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java:18]: Specifically it comes from the Spring env, which includes application.yml properties. If it's present in the application.yml it should still get picked up by ConditionalOnProperty

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java [hunk 16]
```diff
@@ -27,7 +27,7 @@ public class BootstrapManagerFactory {
   @Qualifier("entityService")
   private EntityService _entityService;
 
-  @Autowired
+  @Autowired(required = false)
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java:30]: Yeah it should. Each step should have its own factory but nonetheless the step shouldn't be depending on Ebean specifically. This seems dangerous

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java:18]: but in the application.yml we should be doing something like

```
entityService:
    impl: 'ebean' 
    config
....
```

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (RyanHolstien) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java:30]: @Autowired(required = false) basically just makes it not force wiring, if it's there it's there, if it's not then it's usually accompanied by null checks/applicationContext checks. Since this class doesn't actually use the object, I don't think this should block the PR for another day as it should not cause any problems, but we should probably just remove it fully after this goes in.

[the remark this answers is not in the document]
Comment (RyanHolstien) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java:30]: Missed it getting passed into the constructor here, yeah this will cause that step to fail if it's not present. We should pull this functionality out directly into the EntityService rather than have this step directly accessing the EbeanServer.

[the remark this answers is not in the document]
Comment (justinas-marozas) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java:30]: Yeah, it would be best if it just depended on the EntityService. Should it be sorted in this PR?

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (justinas-marozas) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/entity/EbeanAspectDaoFactory.java]: Added entityService.impl and cassandra config to the application.yml

**Review by jjoyce0510 — APPROVED** [review]
LGTM - Thank you guys for all the hard work.

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/factories/BootstrapManagerFactory.java:30]: We have decided to address this on our side in a followup. No need to churn you kind folks on it.
